cmd/kubectl-directpv/drives_move.go

```diff
@@ -18,133 +18,161 @@ package main
-		klog.Error("the source and target drives does not belong to the same access-tier")
-		return errInvalidState
+
+	if srcDrive.GetAccessTier() != destDrive.GetAccessTier() {
+		eprintf(
+			fmt.Sprintf(
+				"source drive access-tier %v and destination drive access-tier %v are different",
+				srcDrive.GetAccessTier(),
+				destDrive.GetAccessTier(),
+			),
+			true,
+		)
+		os.Exit(1)
 	}
-	return nil
-}
 
-func move(ctx context.Context, sourceDrive, targetDrive *types.Drive) error {
-	selector, err := getDriveNameSelectors([]string{sourceDrive.Name})
-	if err != nil {
-		return err
+	if destDrive.Status.FreeCapacity < requiredCapacity {
+		eprintf(
+			fmt.Sprintf(
+				"insufficient free capacity on destination drive; required=%v free=%v",
+				destDrive.Name,
+				printableBytes(requiredCapacity),
+				printableBytes(destDrive.Status.FreeCapacity),
+			),
+			true,
+		)
+		os.Exit(1)
 	}
-	volumes, err := volume.GetVolumeList(ctx, nil, nil, nil, nil, selector)
+
+	destDrive.Status.FreeCapacity -= requiredCapacity
+	destDrive.Status.AllocatedCapacity += requiredCapacity
+	for _, volume := range volumes {
+		destDrive.AddVolumeFinalizer(volume)
+	}
+	destDrive.Status.Status = directpvtypes.DriveStatusMoving
+	_, err = client.DriveClient().Update(
+		ctx, destDrive, metav1.UpdateOptions{TypeMeta: types.NewDriveTypeMeta()},
+	)
 	if err != nil {
-		return err
+		eprintf(fmt.Sprintf("unable to move volumes to destination drive; %v", err), true)
+		os.Exit(1)
 	}
+
 	for _, volume := range volumes {
-		if volume.Status.DriveName != sourceDrive.Name {
-			klog.Infof("invalid drive name %s found in volume %s", volume.Status.DriveName, volume.Name)
-			return errInvalidState
-		}
-		if volume.Status.NodeName != sourceDrive.Status.NodeName {
-			klog.Infof("invalid node name %s found in volume %s", volume.Status.NodeName, volume.Name)
-			return errInvalidState
-		}
-		if volume.IsPublished() {
-			klog.Info("please make sure all the volumes in the source drive are not inuse")
-			return fmt.Errorf("volume %s is still published", volume.Name)
-		}
-		if targetDrive.Status.FreeCapacity < volume.Status.TotalCapacity {
-			klog.Info("the target drive cannot accomodate the volumes from the source")
-			return errCapacityNotAvailable
-		}
-		finalizer := consts.DriveFinalizerPrefix + volume.Name
-		if !utils.ItemIn(targetDrive.Finalizers, finalizer) {
-			targetDrive.Status.FreeCapacity -= volume.Status.TotalCapacity
-			targetDrive.Status.AllocatedCapacity += volume.Status.TotalCapacity
-			targetDrive.SetFinalizers(append(targetDrive.GetFinalizers(), finalizer))
-		}
+		eprintf(fmt.Sprintf("Moving volume %v", volume), false)
+	}
+
+	srcDrive.ResetFinalizers()
+	_, err = client.DriveClient().Update(
+		ctx, srcDrive, metav1.UpdateOptions{TypeMeta: types.NewDriveTypeMeta()},
+	)
+	if err != nil {
+		eprintf(fmt.Sprintf("unable to remove volume references in source drive; %v", err), true)
+		os.Exit(1)
 	}
```

Comment:
we don't need to reset the finalizers of the source drive, we can still keep it for reference. `drives release` can handle these moved drives by resetting the finalizers and updating

Also, this function will not be idempotent if we bring two updates, ie, if the second update fails, the retry will not be idempotent. ie, if the target drive is already in moving state, retrying it will fail.